src/libsync/vfs/cfapi/vfs_cfapi.cpp

```diff
@@ -269,17 +269,39 @@ Vfs::AvailabilityResult VfsCfApi::availability(const QString &folderPath)
     return AvailabilityError::NoSuchItem;
 }
 
-void VfsCfApi::cancelHydration(const QString &requestId, const QString & /*path*/)
+HydrationJob *VfsCfApi::findHydrationJob(const QString &requestId) const
 {
     // Find matching hydration job for request id
     const auto hydrationJobsIter = std::find_if(d->hydrationJobs.cbegin(), d->hydrationJobs.cend(), [&](const HydrationJob *job) {
         return job->requestId() == requestId;
     });
 
-    // If found, cancel it
     if (hydrationJobsIter != d->hydrationJobs.cend()) {
-        (*hydrationJobsIter)->cancel();
+        return *hydrationJobsIter;
+    }
+
+    return nullptr;
+}
+
+void VfsCfApi::cancelHydration(const QString &requestId, const QString & /*path*/)
+{
+    // Find matching hydration job for request id
+    const auto hydrationJob = findHydrationJob(requestId);
+    // If found, cancel it
+    if (hydrationJob) {
+        qCInfo(lcCfApi) << "Cancel hydration";
+        hydrationJob->cancel();
+    }
+}
+
+bool VfsCfApi::hydrationCancelled(const QString &requestId) const
```

Comment:
if I understand correctly calls to `hydrationCancelled` are coming from callbacks of CfApi
that means that they come from different threads and accesses within `findHydrationJob` can lead to data races

Reply:
so there are two options here:

- we add mutex to ensure data races cannot happen
- we switch to a lock free solution by modifying the communication channel to use for example another local socket such that we have two socket, one for the data and one for getting cancelation to work

Reply:
done